File: UserSpace/Cpp/Axion/Axion.cpp
#include <iostream>
#include <iomanip> 
#include <cmath> 
#include <string> 

#include"MiMeS.hpp"

// #define printPoints // print the entire evolution of the axion angle, its derivative, and other quantities
// #define printPeaks // print these quantities at the peaks of the oscillation
// #define printError //print only the results
// #define smallPrint //print the results in a human readable form
#define resultPrint //print only the results


// macros for the solver
#ifndef SOLVER
    #define SOLVER 1
    #define METHOD RODASPR2
#endif


// macros for the numeric type
#ifndef LONG
    #define LONG 
#endif

#ifndef LD
    #define LD LONG double
#endif

int main(int argc, char **argv){ 

    if(argc!=18){
        std::cerr<<"usage: Axion.run theta_i f_a umax TSTOP ratio_ini N_convergence_max convergence_lim inputFile\n \
        initial_step_size minimum_step_size maximum_step_size absolute_tolerance relative_tolerance beta \n \
        fac_max fac_min maximum_No_steps\n";
        
        std::cerr<<"With:\n\n \
        theta_i: initial angle\n\n \
        fa: PQ scale in GeV (the temperature dependent mass is defined as m_a^2(T) = \\chi(T)/f^2)\n\n \
        umax: if u>umax the integration stops (rempember that u=log(a/a_i))\n\n \
        TSTOP: if the temperature drops below this, integration stops\n\n \
        ratio_ini: integration starts when 3H/m_a<~ratio_ini (this is passed to AxionEOM,\n \
        in order to make the interpolations start at this point)\n\n \
        N_convergence_max and convergence_lim: integration stops when the relative difference\n \
        between two consecutive peaks is less than convergence_lim for N_convergence_max\n \
        consecutive peaks\n\n \
        inputFile: file that describes the cosmology. the columns should be: u T[GeV] logH\n\n \
        -----------Optional arguments------------------------\n\n \
        initial_stepsize: initial step the solver takes.\n\n \
        maximum_stepsize: This limits the sepsize to an upper limit.\n\n \
        minimum_stepsize: This limits the sepsize to a lower limit.\n\n \
        absolute_tolerance: absolute tolerance of the RK solver.\n\n \
        relative_tolerance: relative tolerance of the RK solver.\n\n \
        Note:\n \
        Generally, both absolute and relative tolerances should be 1e-8.\n\n \
        In some cases, however, one may need more accurate result (eg if f_a is extremely high,\n \
        the oscillations happen violently, and the ODE destabilizes). Whatever the case, if the\n \
        tolerances are below 1e-8, long doubles *must* be used.\n\n \
        beta: controls how agreesive the adaptation is. Generally, it should be around but less than 1.\n\n \
        fac_max,  fac_min: the stepsize does not increase more than fac_max, and less than fac_min.\n\n \
        This ensures a better stability. Ideally, fac_max=inf and fac_min=0, but in reality one must\n \
        tweak them in order to avoid instabilities.\n\n \
        maximum_No_steps: maximum steps the solver can take Quits if this number is reached even if integration\n \
        is not finished.\n";

        return 1;
    }

    
    //timeit.sh adds some (insignificant) overhead. So, alternatively, you can use
    // mimes::util::Timer _time_; 
    // The _time_ instance  prints the time the program took in stderr when the program exits (when _time_ goes out of scope in general).  

    // you can change the axion mass function for T above the interpolation range like this (these are not correct; they only show how it can be done):
    // axionMass<LD>.ma2_MAX = [](LD T, LD fa)->LD{ return 0; }; // this is fine for T>>T_{QCD}
    // axionMass<LD>.ma2_MIN = [](LD T, LD fa)->LD{return 0;}; //this gives Omegah^2 = 0 (but you are sure that it works)

    int ar=0;

    //model parameters
    LD theta_i = atof(argv[++ar]) ;
    LD fa = atof(argv[++ar]);
    // solver parameters
    LD umax = atof(argv[++ar]); //u at which the integration stops 
    LD TSTOP = atof(argv[++ar]); // temperature at which integration stops
    LD ratio_ini=atof(argv[++ar]); // 3H/m_a at which integration begins (should be larger than 500 or so)
    // stopping conditions.
    // integration stops after the adiabatic invariant hasn't changed 
    // more than  convergence_lim% for N_convergence_max consecutive peaks
    unsigned int N_convergence_max=atoi(argv[++ar]);
    LD convergence_lim=atof(argv[++ar]);
    //file in which the cosmology is defined. the columns should be : u T[GeV] logH
    std::string  inputFile=argv[++ar];
    
    /*options for the solver (These variables are optional. Yoou can use the Axion class without them).*/
    LD initial_step_size=atof(argv[++ar]); //initial step the solver takes. 
    LD minimum_step_size=atof(argv[++ar]); //This limits the sepsize to an upper limit. 
    LD maximum_step_size=atof(argv[++ar]); //This limits the sepsize to a lower limit.
    LD absolute_tolerance=atof(argv[++ar]); //absolute tolerance of the RK solver
    LD relative_tolerance=atof(argv[++ar]); //relative tolerance of the RK solver
    LD beta=atof(argv[++ar]); //controls how agreesive the adaptation is. Generally, it should be around but less than 1.
    /*
    the stepsize does not increase more than fac_max, and less than fac_min. 
    This ensures a better stability. Ideally, fac_max=inf and fac_min=0, but in reality one must 
    tweak them in order to avoid instabilities.
    */
    LD fac_max=atof(argv[++ar]); 
    LD fac_min=atof(argv[++ar]);
    unsigned int maximum_No_steps=atoi(argv[++ar]); //maximum steps the solver can take Quits if this number is reached even if integration is not finished.
    

    // use chi_PATH to interpolate the axion mass.
    mimes::AxionMass<LD> axionMass(chi_PATH,0,mimes::Cosmo<LD>::mP);
    /*set ma2 for T>TMax*/
    LD TMax=axionMass.getTMax();    
    LD chiMax=axionMass.getChiMax();    
    axionMass.set_ma2_MAX( [&chiMax,&TMax](LD T, LD fa){return chiMax/fa/fa*std::pow(T/TMax,-8.16);});  
    
    /*set ma2 for T<TMin*/
    LD TMin=axionMass.getTMin();  
    LD chiMin=axionMass.getChiMin();    
    axionMass.set_ma2_MIN( [&chiMin,&TMin](LD T, LD fa){return chiMin/fa/fa;});

    // you can define the axion mass using a function, like this 
    // std::function<LD(LD,LD)> ma2 = [](LD T,LD fa){
    //     LD TQCD=150*1e-3;
    //     LD ma20=3.1575e-05/fa/fa;
    //     if(T<=TQCD){return ma20;}
    //     else{return ma20*std::pow((TQCD/T),8.16);}
    // };
    // mimes::AxionMass<LD> axionMass(ma2);


    // instance of Axion
    mimes::Axion<LD,SOLVER,METHOD<LD>> ax(theta_i, fa, umax, TSTOP, ratio_ini, N_convergence_max,
    convergence_lim, inputFile, &axionMass,
    initial_step_size,minimum_step_size, maximum_step_size, absolute_tolerance, relative_tolerance, beta,
    fac_max, fac_min, maximum_No_steps);

    // Solve the Axion EOM
    ax.solveAxion();

    //get the most important results 
    #ifdef resultPrint
    std::cout<<std::setprecision(16)
    <<theta_i<<" "<< fa<<" "<<ax.theta_osc<<" "<<ax.T_osc<<" "<<ax.relic<<"\n";
    #endif 

    #ifdef smallPrint
    std::cout<<std::setprecision(25)
    <<"theta_i="<<theta_i<<" "<<"f_a="<< fa<<" GeV\n"<<"theta_osc~="<<ax.theta_osc<<" "
    <<"T_osc~="<<ax.T_osc<<"GeV \n"
    <<"Omega h^2="<<ax.relic<<"\n";
    #endif 


    // print all the points
    #ifdef printPoints
    std::cout<<"---------------------points:---------------------\n";
    std::cout<<"a/a_i\tT [GeV]\ttheta\tzeta\trho_a [GeV^4]"<<std::endl;
    for(int i=0; i<ax.pointSize; ++i ){
        for(int j=0; j<5; ++j){
            std::cout<<ax.points[i][j];
            if(j==4){std::cout<<"\n";}else{std::cout<<"\t";}
        }
    }
    #endif

    // print all the peaks
    #ifdef printPeaks
    std::cout<<"---------------------peaks:---------------------\n";
    std::cout<<"a/a_i\tT [GeV]\ttheta\tzeta\trho_a [GeV^4]\tadiabatic_inv [GeV]"<<std::endl;
    for(int i=0; i<ax.peakSize; ++i ){
        for(int j=0; j<6; ++j){
            std::cout<<ax.peaks[i][j];
            if(j==5){std::cout<<"\n";}else{std::cout<<"\t";}
        }
    }
    #endif

    // print the local errors
    #ifdef printError
    std::cout<<"---------------------local errors:---------------------\n";
    std::cout<<"dtheta\tdzeta"<<std::endl;
    for(size_t i=0; i<ax.pointSize; ++i ){
        std::cout<<ax.dtheta[i]<<"\t"<<ax.dzeta[i]<<"\n";
    }
    #endif


    return 0;
}

File: UserSpace/Python/Axion.py
from time import time
from sys import stderr

from sys import path as sysPath
sysPath.append('../../src')

#load the module
from interfacePy.Axion import Axion 
from interfacePy.AxionMass import AxionMass 
from interfacePy.Cosmo import mP


theta_i, fa=0.94435, 1e12

umax=500
TSTOP=1e-4
ratio_ini=1e3

N_convergence_max, convergence_lim=5, 1e-2 #this is fine, but you can experiment a bit. 

#radiation dominated example
inputFile="../InputExamples/RDinput.dat" 

# Matter Domination example. 
# the NSC parameters (using the notation of 2012.07202) are:
# T_end=1e-2 (GeV), c=3, T_ini=1e12 (GeV), and r=1e-1
# inputFile="../InputExamples/MatterInput.dat" 

# Kination Domination example. 
# the NSC parameters (using the notation of 2012.07202) are:
# T_end=0, c=6, T_ini=1e3 (GeV), and r=1e10
# inputFile="../InputExamples/KinationInput.dat" 



# options for the solver
# These variables are optional. Yoou can use the Axion class without them.
initial_step_size=1e-2; #initial step the solver takes. 
minimum_step_size=1e-8; #This limits the sepsize to an upper limit. 
maximum_step_size=1e-2; #This limits the sepsize to a lower limit.
absolute_tolerance=1e-8; #absolute tolerance of the RK solver
relative_tolerance=1e-8; #relative tolerance of the RK solver
beta=0.9; #controls how agreesive the adaptation is. Generally, it should be around but less than 1.

#The stepsize does not increase more than fac_max, and less than fac_min. 
#This ensures a better stability. Ideally, fac_max=inf and fac_min=0, but in reality one must 
#tweak them in order to avoid instabilities.
fac_max=1.2; 
fac_min=0.8;
maximum_No_steps=int(1e7); #maximum steps the solver can take Quits if this number is reached even if integration is not finished.

_=time()
# AxionMass instance
axionMass = AxionMass(r'../../src/data/chi.dat',0,mP)
#------------------------------------------------------------------------------#
# this is the axion mass squared beyond the interpolation limits for the current data 
# if yo don't specify it, the axion mass is taken to be constant beyond these limits
TMax=axionMass.getTMax() 
chiMax=axionMass.getChiMax()
TMin=axionMass.getTMin() 
chiMin=axionMass.getChiMin()

axionMass.set_ma2_MAX( lambda T,fa: chiMax/fa/fa*pow(T/TMax,-8.16) )
axionMass.set_ma2_MIN( lambda T,fa: chiMin/fa/fa )
#------------------------------------------------------------------------------#
# def ma2(T,fa):
#     TQCD=150*1e-3
#     ma20=3.1575e-05/fa/fa
#     if T<=TQCD:
#         return ma20
#     return ma20*pow((TQCD/T),8.16)
# axionMass = AxionMass(ma2)

# Axion instance
ax=Axion(theta_i, fa, umax, TSTOP, ratio_ini, N_convergence_max, convergence_lim, inputFile, axionMass,
        initial_step_size,minimum_step_size, maximum_step_size, absolute_tolerance, 
        relative_tolerance, beta, fac_max, fac_min, maximum_No_steps)


# solve the EOM (this only gives you the relic, T_osc, theta_osc, and a_osc)
ax.solveAxion()

print(ax.theta_i, ax.fa, ax.theta_osc, ax.T_osc ,ax.relic)
print(round(time()-_,3),file=stderr)



# change to True in order to mkae the plots
if False:
    import matplotlib.pyplot as plt
    from numpy import array as np_array
    from numpy import abs as np_abs

    ax.getPeaks()#this gives you the peaks of the oscillation
    ax.getPoints()#this gives you all the points of integration
    ax.getErrors()#this gives you local errors of integration


    #########-----\theta-----#########
    fig=plt.figure(figsize=(9,4))
    fig.subplots_adjust(bottom=0.15, left=0.15, top = 0.9, right=0.9,wspace=0.0,hspace=0.25)
    sub = fig.add_subplot(1,1,1)
    
    #this plot shows the peaks of the oscillation
    sub.plot(ax.T_peak,ax.theta_peak,linestyle=':',marker='+',color='xkcd:blue',linewidth=2)

    #this plot shows all the points
    sub.plot(ax.T,ax.theta,linestyle='-',linewidth=2,alpha=1,c='xkcd:black')

    
    sub.set_yscale('linear')
    sub.set_xscale('linear')
    
    sub.set_xlabel(r'$T ~[{\rm GeV}]$')
    sub.xaxis.set_label_coords(0.5, -0.1) 
    sub.set_ylabel(r'$\theta$')
    sub.yaxis.set_label_coords(-0.1,0.5) 

    sub.axhline(ax.theta_osc,linestyle=':',color='xkcd:red',linewidth=1.5)
    sub.axvline(ax.T_osc,linestyle='--',color='xkcd:gray',linewidth=1.5)
   
    fig.savefig('theta-T_examplePlot.pdf',bbox_inches='tight')



    #########-----\zeta-----#########
    fig=plt.figure(figsize=(9,4))
    fig.subplots_adjust(bottom=0.15, left=0.15, top = 0.9, right=0.9,wspace=0.0,hspace=0.25)
    sub = fig.add_subplot(1,1,1)
    
    sub.plot(ax.T,ax.zeta,linestyle='-',linewidth=2,alpha=1,c='xkcd:black')
    sub.plot(ax.T_peak,ax.zeta_peak,linestyle=':',marker='+',color='xkcd:blue',linewidth=2)
    
    sub.set_yscale('linear')
    sub.set_xscale('linear')
    
    sub.set_xlabel(r'$T ~[{\rm GeV}]$')
    sub.xaxis.set_label_coords(0.5, -0.1) 
    sub.set_ylabel(r'$\zeta $')
    sub.yaxis.set_label_coords(-0.1,0.5) 

    sub.axvline(ax.T_osc,linestyle='--',color='xkcd:gray',linewidth=1.5)
    fig.savefig('zeta-T_examplePlot.pdf',bbox_inches='tight')


    #########-----errors-----#########
    fig=plt.figure(figsize=(9,4))
    fig.subplots_adjust(bottom=0.15, left=0.15, top = 0.9, right=0.9,wspace=0.0,hspace=0.25)
    sub = fig.add_subplot(1,1,1)
    
    sub.plot(ax.T,np_abs(ax.dtheta/ax.theta),linestyle='-',linewidth=2,alpha=1,c='xkcd:black',label=r'$\dfrac{\delta \theta}{\theta}$')
    sub.plot(ax.T,np_abs(ax.dzeta/ax.zeta),linestyle='-',linewidth=2,alpha=1,c='xkcd:red',label=r'$\dfrac{\delta \zeta}{\zeta}$')
    
    sub.set_yscale('log')
    sub.set_xscale('linear')
    
    sub.set_xlabel(r'$T ~[{\rm GeV}]$')
    sub.xaxis.set_label_coords(0.5, -0.1) 
    sub.set_ylabel(r'local errors')
    sub.yaxis.set_label_coords(-0.1,0.5) 
    sub.legend(bbox_to_anchor=(1, 0.0),borderaxespad=0., 
           borderpad=0.05,ncol=1,loc='lower right',fontsize=14,framealpha=0)

    sub.axvline(ax.T_osc,linestyle='--',color='xkcd:gray',linewidth=1.5)
    fig.savefig('errors_examplePlot.pdf',bbox_inches='tight')



    #########-----rho_a-----#########    
    fig=plt.figure(figsize=(9,4))
    fig.subplots_adjust(bottom=0.15, left=0.15, top = 0.9, right=0.9,wspace=0.0,hspace=0.25)
    sub = fig.add_subplot(1,1,1)
    
    sub.plot(ax.T,ax.rho_axion,linestyle='-',linewidth=2,alpha=1,c='xkcd:black')
    sub.plot(ax.T_peak,ax.rho_axion_peak,linestyle=':',linewidth=2,alpha=1,c='xkcd:blue')

    sub.set_yscale('log')
    sub.set_xscale('linear')
    
    sub.set_xlabel(r'$T ~[{\rm GeV}]$')
    sub.xaxis.set_label_coords(0.5, -0.1) 
    sub.set_ylabel(r'$\rho_{a}(T) ~[{\rm GeV}^{4}]$')
    sub.yaxis.set_label_coords(-0.1,0.5) 
    
    
    sub.axvline(ax.T_osc,linestyle='--',color='xkcd:gray',linewidth=1.5)
    fig.savefig('rho_a-T_examplePlot.pdf',bbox_inches='tight')




#don't forget the destructor
del ax